File: test/CSharp/CodeCracker.Test/Usage/UriAnalyzerTests.cs
using System;
using System.Threading.Tasks;
using CodeCracker.CSharp.Usage;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.Testing;
using Xunit;

namespace CodeCracker.Test.CSharp.Usage
{
    public class UriAnalyzerTests : CodeFixVerifier
    {
        private const string TestCode = @"
using System;
namespace ConsoleApplication1
{{
    class Person
    {{
        public Person()
        {{
            {0}
        }}
    }}
}}";

#pragma warning disable CC0063
        [Fact]
        public async Task IfAbbreviatedUriConstructorFoundAndUriIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new Uri(""foo"")");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(9, 31, () => new Uri("foo")));
        }

        [Fact]
        public async Task IfUriConstructorFoundAndUriIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new System.Uri(""foo"")");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(9, 38, () => new Uri("foo")));
        }

        [Fact]
        public async Task IfUriConstructorWithUriKindFoundAndUriIsIncorrectCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new Uri(""http://wrong"", UriKind.Relative)");
            await VerifyCSharpDiagnosticAsync(test, CreateDiagnosticResult(9, 31, () => new Uri("http://wrong", UriKind.Relative)));
        }
#pragma warning restore CC0063

        [Fact]
        public async Task IfAbbreviatedUriConstructorWithUriKindFoundAndUriIsCorrectDoNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new Uri(""foo"", UriKind.Relative)");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IfUriConstructorWithUriKindFoundAndUriIsCorrectDoNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new System.Uri(""foo"", UriKind.Relative)");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IfUriConstructorUsingNullFoundDoNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new System.Uri(null)");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IfUriConstructorNotUsingLiteralFoundDoNotCreatesDiagnostic()
        {
            var test = string.Format(TestCode, @"var uri = new System.Uri(new object().ToString())");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IfAbbreviatedUriConstructorFoundAndUriIsIncorrectAndItsNotSystemUriDoNotCreatesDiagnostic()
        {
            const string code = @"
namespace ConsoleApplication1
{
    class Uri
    {
        public Uri(string uri) { }

        public void Test() {
            var uri = new Uri(""whoCares"");
        }
    }
}";
            await VerifyCSharpHasNoDiagnosticsAsync(code);
        }

        private static DiagnosticResult CreateDiagnosticResult(int line, int column, Action getErrorMessageAction)
        {
            return new DiagnosticResult(DiagnosticId.Uri.ToDiagnosticId(), DiagnosticSeverity.Error)
                .WithLocation(line, column)
                .WithMessage(GetErrorMessage(getErrorMessageAction));
        }

        private static string GetErrorMessage(Action action)
        {
            try
            {
                action.Invoke();
            }
            catch (Exception ex)
            {
                return ex.Message;
            }
            return "";
        }

        protected override DiagnosticAnalyzer GetDiagnosticAnalyzer()
        {
            return new UriAnalyzer();
        }
    }
}

File: test/VisualBasic/CodeCracker.Test/Usage/UriAnalyzerTests.vb
Imports CodeCracker.VisualBasic.Usage
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

Namespace Usage
    Public Class UriAnalyzerTests
        Inherits CodeFixVerifier
        Private Const TestCode As String = "
Imports System
Namespace ConsoleApplication1
	Class Person
		Public Sub New()
            {0}
		End Sub
	End Class
End Namespace
"

#Disable Warning CC0063
        <Fact>
        Public Async Function IfAbbreviatedUriConstructorFoundAndUriIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New Uri(""foo"")")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(6, 31, Function() New Uri("foo")))
        End Function

        <Fact>
        Public Async Function IfUriConstructorFoundAndUriIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New System.Uri(""foo"")")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(6, 38, Function() New Uri("foo")))
        End Function

        <Fact>
        Public Async Function IfUriConstructorWithUriKindFoundAndUriIsIncorrectCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New Uri(""http://wrong"", UriKind.Relative)")
            Await VerifyBasicDiagnosticAsync(test, CreateDiagnosticResult(6, 31, Function() New Uri("http://wrong", UriKind.Relative)))
        End Function
#Enable Warning CC0063

        <Fact>
        Public Async Function IfAbbreviatedUriConstructorWithUriKindFoundAndUriIsCorrectDoNotCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New Uri(""foo"", UriKind.Relative)")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfUriConstructorWithUriKindFoundAndUriIsCorrectDoNotCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New System.Uri(""foo"", UriKind.Relative)")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfUriConstructorUsingNullFoundDoNotCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New System.Uri(null)")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfUriConstructorNotUsingLiteralFoundDoNotCreatesDiagnostic() As Task
            Dim test = String.Format(TestCode, "Dim uri = New System.Uri(new object().ToString())")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IfAbbreviatedUriConstructorFoundAndUriIsIncorrectAndItsNotSystemUriDoNotCreatesDiagnostic() As Task
            Const code As String = "
Namespace ConsoleApplication1
	Class Uri
		Public Sub New(uri__1 As String)
		End Sub

		Public Sub Test()
			Dim uri = New Uri(""whoCares"")
		End Sub
	End Class
End Namespace"
            Await VerifyBasicHasNoDiagnosticsAsync(code)
        End Function

        Private Shared Function CreateDiagnosticResult(line As Integer, column As Integer, getErrorMessageAction As Action) As DiagnosticResult
            Return New DiagnosticResult(DiagnosticId.Uri.ToDiagnosticId(), DiagnosticSeverity.Error) _
                .WithLocation(line, column) _
                .WithMessage(GetErrorMessage(getErrorMessageAction))
        End Function

        Private Shared Function GetErrorMessage(action As Action) As String
            Try
                action.Invoke()
            Catch ex As Exception
                Return ex.Message
            End Try
            Return ""
        End Function

        Protected Overrides Function GetDiagnosticAnalyzer() As DiagnosticAnalyzer
            Return New UriAnalyzer()
        End Function
    End Class
End Namespace